### ISSUE TITLE
Injects successfully, however nothing happens. 

### ISSUE THREAD

**Issue by GBanana** [NodeMixaholic/callienaps-nw#1]
I understand that you probably have left support for this a long time ago, but as a curious individual could you tell me what I would need to update to get this working? Offsets? There's not much information I can find on this topic elsewhere. Thank you.

Mac OS Monterey 
12.2.1

**Comment by NodeMixaholic** [comment]
Honestly, i almost forgot about this project. However, it seems that injection works but executing Lua code doesn't. Sorry for the obscure details, I am bad at C++ and honestly long forgot how to code for this. I just remember that this doesn't work and how it doesn't. Feel free to contribute if you have coding skills. Basically I think I was trying to use a custom lua environment to inject lua code. Note that I developed this with an Intel Mac, it might be different on Apple ARM. Thanks,

NodeMixaholic.

**Comment by GBanana** [comment]
Firstly, thank you for being a great consistent github user, without individuals like you, I wouldn't be able to learn anything at all. Also, I was just wondering how I could get a console log output from injecting this dylib, as this would help me in debugging and finding where the issues lie at.

**Comment by GBanana** [comment]
I also believe this should be easier for me to do now as roblox has made public their version of roblox lua luau.

**Comment by NodeMixaholic** [comment]
They really made LUAU public? ^_^ That's amazing.

**Comment by NodeMixaholic** [comment]
Update: I am now slowly adding/removing LUAU .h files and seeing what breaks. So far, most of them broke my code (which is why I have a GitHub, to keep track of my code in case I need a copy of it.) TLDR: I am hoping by adding as much LUAU as I can, I could hopefully finally get things working.

**Comment by NodeMixaholic** [comment]
> Firstly, thank you for being a great consistent github user, without individuals like you, I wouldn't be able to learn anything at all. Also, I was just wondering how I could get a console log output from injecting this dylib, as this would help me in debugging and finding where the issues lie at.

First off, thank you for your support. It is with great pleasure I introduce a patch where, if it successfully works (and thats a big if, because this is a hobby and nothing big like synapse), it will print to the ROBLOX console that it works. And not only that, but it will also pop open a GUI. Again, IF it works. Thanks again for your support, I really am a n00b when it comes to ROBLOX hacks, and it is truly great doing this, especially when people 
 like you support me.

**Comment by GBanana** [comment]
No, it's great that you've made this project open sourced, something that not many people would do at all, I have a few questions regarding your code, you have an address in rewrite.cpp as luaustringaddress = 0x0...
What address is this linked to? Also, when you said you introduced a patch with great pleasure, does that mean you got the code working? Or was this just a referring to a successful compile after adding luau code into the project? I spent a while in XCode adding and removing lua libraries, but I can't get it to work with the built in GNU compiler, but I wouldn't worry about that since when using terminal it compiles perfectly. As for my support, it is very little, and it's the least I can do for free education.

**Comment by GBanana** [comment]
Just before I sleep if anyone else is here reading this, the dylib has an error of "SEGMENTATION FAULT 11", I'm assuming this is due to it not being injected, but I can't access the injected version, I tried using DYLYD_PRINT_APIS to get errors but it doesn't work, I believe new instances of the library are being created each time.

**Comment by NodeMixaholic** [comment]
> No, it's great that you've made this project open sourced, something that not many people would do at all, I have a few questions regarding your code, you have an address in rewrite.cpp as luaustringaddress = 0x0...
> 
> What address is this linked to? Also, when you said you introduced a patch with great pleasure, does that mean you got the code working? Or was this just a referring to a successful compile after adding luau code into the project? I spent a while in XCode adding and removing lua libraries, but I can't get it to work with the built in GNU compiler, but I wouldn't worry about that since when using terminal it compiles perfectly. As for my support, it is very little, and it's the least I can do for free education.

By "luaustringaddress" I assume you mean my modified version of the first few lines of: https://github.com/Speedstarskiwi/LuaU-Extensions .  See there for details about that. Secondly, this doesn't work with the default compiler, as we need certain flags and instances for it to work not provided by default. See instructions about "brew", which I assume you already did because you got it to compile via terminal OK. Thanks once again for you support, every little bit helps. ^_^

PS: I can't reveal much about the secret soup I am trying to make without it being cloned, but LUA sits atop LUAU to bring the old LUA libs to the new, LUA is necessary.

**Comment by NodeMixaholic** [comment]
Also, no the hacks don't work yet (i don't think)

**Comment by GBanana** [comment]
I've been doing some stuff lately, and I found that the program, is not executing in roblox. I haven't been able to come up with a reason as to why yet, but it's not executing within the roblox process once executed. The proof I have of this is that I created a simple script within int main() where a text file is modified by the dylib once injected. When the dylib is injected within roblox, nothing occurs. To further prove this point I debugged the program in XCode and the int main() function run successfully and the text file did change, however the dylib program is then halted by an error of "Thread 1: EXC_BAD_ACCESS (code=1, address=0x78)" which occurs right after it console logs "Loading Lua....", I used to think that this only occurred since I was debugging in XCode, and the dylib wasn't actually injected within roblox, thus the address is invalid. But now after I learnt that the text file isn't being modified, it's safe to say that the dylib's code isn't actually being run within roblox. The weird part is how with the older version of callienaps when you run the dylib, you receive a finder prompt saying "Roblox would like to access your X folder" with x being the folder of callienaps, so I assume the older version DID run successfully within roblox and called for the gui.lua and customluafunc.lua and the error occured AFTER it called for those.

Also sorry for not using paragraphing it's probably quite hard to read.

TL;DR = Latest version of Dylib is not running when injected into roblox at all. Cause: Unknown however the 
Older version of dylib is running when injected into roblox.

**Comment by GBanana** [comment]
Alr I take most of that back, the text file itself is not related to the dylibs activity within roblox, but what I do know is that the older file does ask for folder access permissions unlike the newer file, which might mean something? Not sure rlly, debugging the older file stops it after it logs "Loading loadstrings..." at the Lua_pushclosure function

**Comment by NodeMixaholic** [comment]
> Alr I take most of that back, the text file itself is not related to the dylibs activity within roblox, but what I do know is that the older file does ask for folder access permissions unlike the newer file, which might mean something? Not sure rlly, debugging the older file stops it after it logs "Loading loadstrings..." at the Lua_pushclosure function

If you got asked for perms, you got it to work. Congrats. Perhaps you could tell me which version of dylib you chose, and what you changed in the code to get it to work? Thanks so much! :partying_face:

**Comment by NodeMixaholic** [comment]
> > Alr I take most of that back, the text file itself is not related to the dylibs activity within roblox, but what I do know is that the older file does ask for folder access permissions unlike the newer file, which might mean something? Not sure rlly, debugging the older file stops it after it logs "Loading loadstrings..." at the Lua_pushclosure function
> 
> If you got asked for perms, you got it to work. Congrats. Perhaps you could tell me which version of dylib you chose, and what you changed in the code to get it to work? Thanks so much! :partying_face: 

By work I mean you got it to run, not necessarily execute scripts properly. But who knows? It might.

**Comment by NodeMixaholic** [comment]
Update: Just if you have the time, send over a commit to the main branch with your changed Mac compile command. 

"Thanks for helping me help you to help us all." --GLaDOS

**Comment by GBanana** [comment]
Hey there, it's just before midnight right now for me, I'll attend this tmrw morning, but what I have been doing is using xcode's debugger to see where the error occurs. The program stops at the "lua_newthread(L);" within the debugger, this might be just because I'm using xcode's debugger and it's not actually in roblox. I've been trying to play with other injection methods related to killing roblox's pid as soon as it starts, and rebooting it with DYLD_INSERT_LIBRARIES including the callie dylib, but I'm still being greeted by the segmentation fault 11, and I'm literally unable to tell if it even executes or not. I will try again with a different folder path tmrw, using the older dylib file to see if it asks for permissions, which would mean the injection method works, if it doesn't then I just wasted another day doing random things. But it's fine since this is fun for me.

**Comment by NodeMixaholic** [comment]
> Hey there, it's just before midnight right now for me, I'll attend this tmrw morning, but what I have been doing is using xcode's debugger to see where the error occurs. The program stops at the "lua_newthread(L);" within the debugger, this might be just because I'm using xcode's debugger and it's not actually in roblox. I've been trying to play with other injection methods related to killing roblox's pid as soon as it starts, and rebooting it with DYLD_INSERT_LIBRARIES including the callie dylib, but I'm still being greeted by the segmentation fault 11, and I'm literally unable to tell if it even executes or not. I will try again with a different folder path tmrw, using the older dylib file to see if it asks for permissions, which would mean the injection method works, if it doesn't then I just wasted another day doing random things. But it's fine since this is fun for me. 
> 
> 

First off, have great dreams. Second off, there are a few points I would like to make:

* XCode does not fit our needs, hence we use a brew version of g++.
* Even if it does, I would still use cli because XCode is hard for me to use compared to CLI.
* You are getting the error likely because you are compiling within Xcode without the lua libs. Try importing those.
* Note that I don't officially support XC debug. I am just trying to help.
* I love doing random things to make code work too. Glad I found a dying breed. A true non-corperate developer.

**Comment by GBanana** [comment]
[callienaps-nw-master.zip](https://github.com/NodeMixaholic/callienaps-nw/files/9484708/callienaps-nw-master.zip)
I am unable to get this to reproduce the terminal prompt but it was this version ^^. 

My edits are mainly to find the errors within the dylib but it's given me nothing but more confusion. When the dylibs launch without the aid of lib2proc we get a segmentation fault 11, that occurs right when it executes lua_newthread(L); we can blame this on the fact that it's not injected within roblox, however once the dylibs injected into roblox, my temporary debugging method of writing what step the dylib is on to a file doesn't work. 

I blame it on either A) the script is not executing once it's injected at all. B)The script is injecting, however from where it is within memory it is unable to write to the file I am asking it to write to. 

Also I forgot to tell you, the luaU extensions is using an outdated address, as roblox updates these things nearly every week. This could be to blame for the reason the NEW version doesn't execute, as for the old one, I haven't tried my "debugging" method on it, which is just writing to file. For the correct address I will need to install ida to reverse engineer roblox and I deleted it recently, I am not very experienced with IDA so I'll need to learn a few things, which is why I'm here in the first place.

This is a hobby for me, nothing more, nothing less. 

However I still don't believe that this segmentation fault 11 should be happening, as any other dylib would usually say something like "Cannot execute binary file.". As for me regarding editing the code, about a month ago I knew nothing about this code at all, and I wanted to know what each thing meant in it, now I can explain everything that's written in the code, the only real edits I've made to the code were probably simplifying some of the .sh files and adding a function in rewrite.cpp that writes to a txt file to know which step the code errors in (However even this doesn't work once injected). 

One thing I don't like about lib2proc is the inability to receive output from the dylib, which is why I explored DYLD_INSERT_LIBRARIES, however this results in a segmentation fault 11, as it creates another instance of roblox instead of using the existing roblox instance. 

What is your opinion on remaking the entire CLVM with majority Luau and only a very few amount of essential lua files?

I am a newbie and I recognise that, this is a learning experience for me, if I can help you to get this to work that's a great bonus, however you would be the captain of the ship.

Also I have solely focused on the dylib, I have completely ignored the DLL as a possibility for exploitation. I am attempting to convey all of my knowledge to this thread regarding this topic so that it can help you, me, or maybe even someone else in a few months.

As per my knowledge a LVM is one of the hardest methods of execution, so patience and dedication is likely key. Aside from this, I have been working on a hook function dylib and it's not going well in terms of bypassing memcheck as roblox crashes instantly, and to hook the function I'd need to update the address I'm hooking too, which would need me to install ida again. 

For anyone that might read this, and is interested in VM's I found this source to be useful:
https://github.com/the0d3v/LoukaPDF/blob/master/OVM.pdf
There is also extra information regarding other types of execution. 

And yes XCode is quite annoying and barely works, most of my experience with it was a waste of time so far, but another learning experience!

Thanks again! Sorry for the ultra long message with no TLDR :(.

**Comment by NodeMixaholic** [comment]
> [callienaps-nw-master.zip](https://github.com/NodeMixaholic/callienaps-nw/files/9484708/callienaps-nw-master.zip)
> 
> I am unable to get this to reproduce the terminal prompt but it was this version ^^. 
> 
> 
> 
> My edits are mainly to find the errors within the dylib but it's given me nothing but more confusion. When the dylibs launch without the aid of lib2proc we get a segmentation fault 11, that occurs right when it executes lua_newthread(L); we can blame this on the fact that it's not injected within roblox, however once the dylibs injected into roblox, my temporary debugging method of writing what step the dylib is on to a file doesn't work. 
> 
> 
> 
> I blame it on either A) the script is not executing once it's injected at all. B)The script is injecting, however from where it is within memory it is unable to write to the file I am asking it to write to. 
> 
> 
> 
> Also I forgot to tell you, the luaU extensions is using an outdated address, as roblox updates these things nearly every week. This could be to blame for the reason the NEW version doesn't execute, as for the old one, I haven't tried my "debugging" method on it, which is just writing to file. For the correct address I will need to install ida to reverse engineer roblox and I deleted it recently, I am not very experienced with IDA so I'll need to learn a few things, which is why I'm here in the first place.
> 
> 
> 
> This is a hobby for me, nothing more, nothing less. 
> 
> 
> 
> However I still don't believe that this segmentation fault 11 should be happening, as any other dylib would usually say something like "Cannot execute binary file.". As for me regarding editing the code, about a month ago I knew nothing about this code at all, and I wanted to know what each thing meant in it, now I can explain everything that's written in the code, the only real edits I've made to the code were probably simplifying some of the .sh files and adding a function in rewrite.cpp that writes to a txt file to know which step the code errors in (However even this doesn't work once injected). 
> 
> 
> 
> One thing I don't like about lib2proc is the inability to receive output from the dylib, which is why I explored DYLD_INSERT_LIBRARIES, however this results in a segmentation fault 11, as it creates another instance of roblox instead of using the existing roblox instance. 
> 
> 
> 
> What is your opinion on remaking the entire CLVM with majority Luau and only a very few amount of essential lua files?
> 
> 
> 
> I am a newbie and I recognise that, this is a learning experience for me, if I can help you to get this to work that's a great bonus, however you would be the captain of the ship.
> 
> 
> 
> Also I have solely focused on the dylib, I have completely ignored the DLL as a possibility for exploitation. I am attempting to convey all of my knowledge to this thread regarding this topic so that it can help you, me, or maybe even someone else in a few months.
> 
> 
> 
> As per my knowledge a LVM is one of the hardest methods of execution, so patience and dedication is likely key. Aside from this, I have been working on a hook function dylib and it's not going well in terms of bypassing memcheck as roblox crashes instantly, and to hook the function I'd need to update the address I'm hooking too, which would need me to install ida again. 
> 
> 
> 
> For anyone that might read this, and is interested in VM's I found this source to be useful:
> 
> https://github.com/the0d3v/LoukaPDF/blob/master/OVM.pdf
> 
> There is also extra information regarding other types of execution. 
> 
> 
> 
> And yes XCode is quite annoying and barely works, most of my experience with it was a waste of time so far, but another learning experience!
> 
> 
> 
> Thanks again! Sorry for the ultra long message with no TLDR :(.

Two words: Thank you. In for words: Thank you so much! :smile:

**Comment by NodeMixaholic** [comment]
> [callienaps-nw-master.zip](https://github.com/NodeMixaholic/callienaps-nw/files/9484708/callienaps-nw-master.zip) I am unable to get this to reproduce the terminal prompt but it was this version ^^.
> 
> My edits are mainly to find the errors within the dylib but it's given me nothing but more confusion. When the dylibs launch without the aid of lib2proc we get a segmentation fault 11, that occurs right when it executes lua_newthread(L); we can blame this on the fact that it's not injected within roblox, however once the dylibs injected into roblox, my temporary debugging method of writing what step the dylib is on to a file doesn't work.
> 
> I blame it on either A) the script is not executing once it's injected at all. B)The script is injecting, however from where it is within memory it is unable to write to the file I am asking it to write to.
> 
> Also I forgot to tell you, the luaU extensions is using an outdated address, as roblox updates these things nearly every week. This could be to blame for the reason the NEW version doesn't execute, as for the old one, I haven't tried my "debugging" method on it, which is just writing to file. For the correct address I will need to install ida to reverse engineer roblox and I deleted it recently, I am not very experienced with IDA so I'll need to learn a few things, which is why I'm here in the first place.
> 
> This is a hobby for me, nothing more, nothing less.
> 
> However I still don't believe that this segmentation fault 11 should be happening, as any other dylib would usually say something like "Cannot execute binary file.". As for me regarding editing the code, about a month ago I knew nothing about this code at all, and I wanted to know what each thing meant in it, now I can explain everything that's written in the code, the only real edits I've made to the code were probably simplifying some of the .sh files and adding a function in rewrite.cpp that writes to a txt file to know which step the code errors in (However even this doesn't work once injected).
> 
> One thing I don't like about lib2proc is the inability to receive output from the dylib, which is why I explored DYLD_INSERT_LIBRARIES, however this results in a segmentation fault 11, as it creates another instance of roblox instead of using the existing roblox instance.
> 
> What is your opinion on remaking the entire CLVM with majority Luau and only a very few amount of essential lua files?
> 
> I am a newbie and I recognise that, this is a learning experience for me, if I can help you to get this to work that's a great bonus, however you would be the captain of the ship.
> 
> Also I have solely focused on the dylib, I have completely ignored the DLL as a possibility for exploitation. I am attempting to convey all of my knowledge to this thread regarding this topic so that it can help you, me, or maybe even someone else in a few months.
> 
> As per my knowledge a LVM is one of the hardest methods of execution, so patience and dedication is likely key. Aside from this, I have been working on a hook function dylib and it's not going well in terms of bypassing memcheck as roblox crashes instantly, and to hook the function I'd need to update the address I'm hooking too, which would need me to install ida again.
> 
> For anyone that might read this, and is interested in VM's I found this source to be useful: https://github.com/the0d3v/LoukaPDF/blob/master/OVM.pdf There is also extra information regarding other types of execution.
> 
> And yes XCode is quite annoying and barely works, most of my experience with it was a waste of time so far, but another learning experience!
> 
> Thanks again! Sorry for the ultra long message with no TLDR :(.


"What is your opinion on remaking the entire CLVM with majority Luau and only a very few amount of essential lua files?" may or may not have been doing that

**Comment by GBanana** [comment]
That's great! The most painful part of this process is that there's no way to get console logs of the error that is occurring once it's injected into roblox, a way for the dylib to communicate outside of roblox, while injected. If there was a way to find what the error was, we could solve it, but right now it's just either: Dylib Works/Dylib doesn't Work

**Comment by kitfildom** [comment]
Hi there, is there still any progress being made?

**Comment by NodeMixaholic** [comment]
i am waiting to see how effective byfron is on macs...
